portable/MemMang/heap_1.c

```diff
@@ -58,7 +58,7 @@
 
 /* The application writer has already defined the array used for the RTOS
 * heap - probably so it can be placed in a special segment or address. */
-    extern uint8_t ucHeap[ configTOTAL_HEAP_SIZE ];
+    extern uint8_t ucHeap[  ];

```

Comment:
That would work for the implementation of heap_4.c, since `ucHeap` is only ever accessed like a pointer would:

https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/6a84f2c1da0f8d9fd1946caa55a336ebacfeb392/portable/MemMang/heap_4.c#L351-L361

In heap_1.c this would break, since the existing implementation accesses `ucHeap[ portBYTE_ALIGNMENT ]`, which would be invalid if `ucHeap` would be a pointer:

https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/6a84f2c1da0f8d9fd1946caa55a336ebacfeb392/portable/MemMang/heap_1.c#L96-L100

Since [pointers and arrays aren't equivalent](https://eli.thegreenplace.net/2009/10/21/are-pointers-and-arrays-equivalent-in-c) I would advise against having the type of ucHeap depend on configAPPLICATION_ALLOCATED_HEAP.

Reply:
Hi @Finwood,

The pointer indexing shouldn't be a problem here. But we are considering how to preperly handle configTOTAL_HEAP_SIZE things. The heap_4 uses configTOTAL_HEAP_SIZE in the init function. The current code force user to set configTOTAL_HEAP_SIZE. However with this change, it may cause some problem if user forget setting configTOTAL_HEAP_SIZE when allocate the heap memory in application. As for now please just made the change from you code to resolve your issue and we will keep you updated on this PR.

Thank you!

Regards,
Ming

Reply:
Changed as requested 👍